### PR TITLE
Charlesmchen/2fa

### PR DESCRIPTION
**Pull request by charlesmchen**
A first draft of the 2FA settings views.

In particular the copy is not final.

PTAL @michaelkirk 

simulator screen shot - iphone 8 plus - 2018-02-28 at 22 39 44
simulator screen shot - iphone 8 plus - 2018-02-28 at 22 42 37
simulator screen shot - iphone 8 plus - 2018-02-28 at 22 42 42
simulator screen shot - iphone 8 plus - 2018-02-28 at 22 42 50
simulator screen shot - iphone 8 plus - 2018-02-28 at 22 50 13

### CODE REVIEW

File: Signal/src/ViewControllers/NotificationSettingsViewController.m [hunk 1]
```diff
@@ -38,21 +38,15 @@ - (void)updateTableContents
...
-        cell.detailTextLabel.text = [OWSSounds displayNameForSound:sound];
-        [cell setAccessoryType:UITableViewCellAccessoryDisclosureIndicator];
-        return cell;
-    }
-                               actionBlock:^{
-                                   OWSSoundSettingsViewController *vc = [OWSSoundSettingsViewController new];
-                                   [weakSelf.navigationController pushViewController:vc animated:YES];
-                               }]];
+    [soundsSection
+        addItem:[OWSTableItem disclosureItemWithText:
```
Comment (charlesmchen) [inline Signal/src/ViewControllers/NotificationSettingsViewController.m]: I DRY'd up some code by adding a new `[OWSTableItem disclosureItemWithText: detailText: actionBlock:]` method for the case where you want "detail text".

File: SignalMessaging/ViewControllers/OWSTableViewController.h [hunk 2]
```diff
@@ -44,11 +44,6 @@ extern const CGFloat kOWSTable_DefaultCellHeight;
 
 #pragma mark -
 
-typedef NS_ENUM(NSInteger, OWSTableItemType) {
```
Comment (charlesmchen) [inline SignalMessaging/ViewControllers/OWSTableViewController.h]: We never ended up needing this.

File: SignalServiceKit/src/Network/API/Requests/OWSRequestFactory.h [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+
+@class TSRequest;
+
+@interface OWSRequestFactory : NSObject
+
+- (instancetype)init NS_UNAVAILABLE;
+
++ (TSRequest *)enable2FARequestWithPin:(NSString *)pin;
+
++ (TSRequest *)disable2FARequest;
```
Comment (charlesmchen) [inline SignalServiceKit/src/Network/API/Requests/OWSRequestFactory.h]: I can't wait to pull more/all of the requests in here.  We might be able to remove 30 files from SSK.

File: SignalServiceKit/src/Network/API/Requests/TSRequest.h [hunk 4]
```diff
@@ -1,11 +1,13 @@
...
-//  Copyright (c) 2017 Open Whisper Systems. All rights reserved.
+//  Copyright (c) 2018 Open Whisper Systems. All rights reserved.
 //
 
 @interface TSRequest : NSMutableURLRequest
 
-@property (nonatomic, retain) NSMutableDictionary *parameters;
+@property (nonatomic) NSDictionary *parameters;
 
-- (void)makeAuthenticatedRequest;
```
Comment (charlesmchen) [inline SignalServiceKit/src/Network/API/Requests/TSRequest.h]: This method was vestigial.

File: SignalServiceKit/src/Network/API/Requests/TSRequest.h [hunk 5]
```diff
@@ -1,11 +1,13 @@
 //
-//  Copyright (c) 2017 Open Whisper Systems. All rights reserved.
+//  Copyright (c) 2018 Open Whisper Systems. All rights reserved.
 //
 
 @interface TSRequest : NSMutableURLRequest
 
-@property (nonatomic, retain) NSMutableDictionary *parameters;
+@property (nonatomic) NSDictionary *parameters;
```
Comment (charlesmchen) [inline SignalServiceKit/src/Network/API/Requests/TSRequest.h]: I'd like to ultimately convert this model to be completely immutable.

File: SignalServiceKit/src/Network/API/Requests/TSRegisterPrekeysRequest.m [hunk 6]
```diff
@@ -17,6 +16,10 @@ - (id)initWithPrekeyArray:(NSArray *)prekeys
        signedPreKeyRecord:(SignedPreKeyRecord *)signedRecord
          preKeyLastResort:(PreKeyRecord *)lastResort {
     self            = [super initWithURL:[NSURL URLWithString:textSecureKeysAPI]];
+    if (!self) {
```
Comment (charlesmchen) [inline SignalServiceKit/src/Network/API/Requests/TSRegisterPrekeysRequest.m]: Baby steps towards consistent init pattern.

File: SignalServiceKit/src/Network/API/Requests/TSRegisterForPushRequest.m [hunk 7]
```diff
@@ -1,29 +1,30 @@
...
+        return nil;
+    }
+
     self.HTTPMethod = @"PUT";
 
-    self.parameters = [@{ @"apnRegistrationId" : identifier } mutableCopy];
+    NSMutableDictionary *parameters = [@{ @"apnRegistrationId" : identifier } mutableCopy];
 
     if (voipId) {
-        [self.parameters setObject:voipId forKeyedSubscript:@"voipRegistrationId"];
```
Comment (michaelkirk) [inline SignalServiceKit/src/Network/API/Requests/TSRegisterForPushRequest.m]: This is some ancient code. I don't think we ever expect this to happen any more. Even on devices that are known not to support voip pushes, like simulators, we return a fake token.

Can we make the voipIdentifier arg non-null and verify nothing else breaks?

We can put this in a card for later if you want.

File: SignalServiceKit/src/Network/API/Requests/OWSRequestFactory.h (hunk 3 above)
Comment (michaelkirk) [reply SignalServiceKit/src/Network/API/Requests/OWSRequestFactory.h]: 🎉

File: Signal/src/ViewControllers/OWS2FASettingsViewController.m [hunk 8]
```diff
@@ -0,0 +1,493 @@
...
+    NSString *right = [oldText substringFromIndex:range.location + range.length].digitsOnly;
+    // 3. Determining the "center" substring: the contents of the new insertion text.
+    NSString *center = insertionText.digitsOnly;
+    // 4. Construct the "raw" new text by concatenating left, center and right.
+    NSString *textAfterChange = [[left stringByAppendingString:center] stringByAppendingString:right];
+    // 5. Ensure we don't exceed the maximum length for a PIN.
+    if (textAfterChange.length > kMaxPinLength) {
+        textAfterChange = [textAfterChange substringToIndex:kMaxPinLength];
+    }
+    // 6. Construct the "formatted" new text by inserting a hyphen if necessary.
```
Comment (michaelkirk) [inline Signal/src/ViewControllers/OWS2FASettingsViewController.m]: I don't see where we're inserting a hyphen. Is this just an old copy/paste comment?

File: Signal/src/ViewControllers/OWS2FASettingsViewController.m [hunk 9]
```diff
@@ -0,0 +1,493 @@
...
+    }
+}
+
+#pragma mark - UITextFieldDelegate
+
+- (BOOL)textField:(UITextField *)textField
+    shouldChangeCharactersInRange:(NSRange)range
+                replacementString:(NSString *)insertionText
+{
+    // TODO: ?
```
Comment (michaelkirk) [inline Signal/src/ViewControllers/OWS2FASettingsViewController.m]: Is the TODO here to make sure we converge with Android's constraints? Could you update the comment?

File: Signal/translations/en.lproj/Localizable.strings [hunk 10]
```diff
@@ -643,6 +643,45 @@
...
 "EMPTY_INBOX_TITLE" = "Squeaky Freaking Clean.";
 
+/* Text field placeholder for 'two factor auth pin' when confirming a pin. */
+"ENABLE_2FA_VIEW_CONFIRM_PIN_DEFAULT_TEXT" = "••••••";
+
+/* Indicates that user should confirm their 'two factor auth pin'. */
+"ENABLE_2FA_VIEW_CONFIRM_PIN_INSTRUCTIONS" = "Confirm your PIN.";
+
+/* Error indicating that attempt to disable 'two-factor auth' failed. */
+"ENABLE_2FA_VIEW_COULD_NOT_DISABLE_2FA" = "Could not disable two-step verification.";
```
Comment (michaelkirk) [inline Signal/translations/en.lproj/Localizable.strings]: I know you said the copy is provisional, but I think it makes sense to use whatever copy we have from the spec as a starting point (e.g. "Registration Lock" not "two-step verification"), hopefully that'll allow us to get more focused feedback when we go over this as a group.

File: Signal/src/ViewControllers/NotificationSettingsViewController.m (hunk 1 above)
Comment (michaelkirk) [reply Signal/src/ViewControllers/NotificationSettingsViewController.m]: 👍

File: Podfile.lock [hunk 11]
```diff
@@ -222,4 +222,4 @@ SPEC CHECKSUMS:
 
 PODFILE CHECKSUM: 66db99df53e7593362ebb004bea5d2215ca00e8e
 
-COCOAPODS: 1.3.1
```
[the remark this answers is not in the document]
Comment (charlesmchen) [reply Podfile.lock]: Done.

File: SignalServiceKit/src/Network/API/Requests/TSRegisterForPushRequest.m (hunk 7 above)
Comment (charlesmchen) [reply SignalServiceKit/src/Network/API/Requests/TSRegisterForPushRequest.m]: Done.

File: Signal/src/ViewControllers/OWS2FASettingsViewController.m (hunk 8 above)
Comment (charlesmchen) [reply Signal/src/ViewControllers/OWS2FASettingsViewController.m]: Done.

File: Signal/src/ViewControllers/OWS2FASettingsViewController.m (hunk 9 above)
Comment (charlesmchen) [reply Signal/src/ViewControllers/OWS2FASettingsViewController.m]: It's around that constant which is fixed in the next PR.

File: Signal/translations/en.lproj/Localizable.strings (hunk 10 above)
Comment (charlesmchen) [reply Signal/translations/en.lproj/Localizable.strings]: Yup, I'm going to fix all of the copy in the next PR.

**Pull request merged**